src/pymovements/datasets/base.py

```diff
@@ -0,0 +1,375 @@
+
+import re
+from abc import ABCMeta
+from abc import abstractmethod
+from pathlib import Path
+from typing import Any
+from urllib.error import URLError
+
+import polars as pl
+from tqdm.auto import tqdm
+
+from pymovements.base import Experiment
+from pymovements.utils.downloads import download_and_extract_archive
+from pymovements.utils.paths import get_filepaths
+
+
+class Dataset:
+    """Dataset base class."""
+
+    def __init__(
+        self,
+        root: str | Path,
+        experiment: Experiment | None = None,
+        filename_regex: str = '.*',
+        filename_regex_dtypes: dict[str, type] | None = None,
+        custom_read_kwargs: dict[str, Any] | None = None,
+    ):
+        """Initialize the dataset object.
+
+        Parameters
+        ----------
+        root : str, Path
+            Path to the root directory of the dataset.
+        experiment : Experiment
+            The experiment definition.
+        filename_regex : str
+            Regular expression which needs to be matched before trying to read the file. Named
+            groups will appear in the `fileinfo` dataframe.
+        filename_regex_dtypes : dict[str, type], optional
+            If named groups are present in the `filename_regex`, this makes it possible to cast
+            specific named groups to a particular datatype.
+        custom_read_kwargs : dict[str, Any], optional
+            If specified, these keyword arguments will be passed to the file reading function.
+        """
+        self.fileinfo = None
+        self.gaze = None
+
+        self.root = Path(root)
+
+        self.experiment = experiment
+
+        self._filename_regex = filename_regex
+
+        if filename_regex_dtypes is None:
+            filename_regex_dtypes = {}
+        self._filename_regex_dtypes = filename_regex_dtypes
+
+        if custom_read_kwargs is None:
+            custom_read_kwargs = {}
+        self._custom_read_kwargs = custom_read_kwargs
+
+    def read(self):
+        """Parse file information and read all gaze files.
+
+        The parsed file information is assigned to the `fileinfo` attribute.
+        All gaze files will be read as dataframes and assigned to the `gaze` attribute.
+        """
+        self.fileinfo = self.read_fileinfo()
+        self.gaze = self.read_gaze_files()
+
+    def read_fileinfo(self) -> pl.DataFrame:
+        """Parse file information from filepaths and filenames.
+
+        Returns
+        -------
+        pl.DataFrame :
+            File information dataframe.
+
+        Raises
+        ------
+        AttributeError
+            If no regular expression for parsing filenames is defined.
+        RuntimeError
+            If an error occured during matching filenames.
+        """
+        if self._filename_regex is not None:
+            filename_regex = re.compile(self._filename_regex)
+        else:
+            raise AttributeError("no regular expression for filenames is defined.")
```

Comment:
this could also be
```python
assert self._filename_regex is not None, "no regular expression for filenames is defined."
filename_regex = re.compil(self._filename_regex)
```

don't know if it might be faster

Reply:
Asserts shouldn't be used in production code as these can get optimized away.

The problem is, that assertions are deactivated when you use the optimization parameter `-O` on launching python. So as an example if a user wants to optimize the interpreter/compiler for memory-efficiency, all of the assertions in this package will be ignored.
Most people don't use this optimization parameter, but I think we should be aware of this issue and maybe even enforce this for newly written code.

Here are some discussions on stackoverflow which can be helpful:
https://stackoverflow.com/questions/944592/best-practice-for-using-assert
https://stackoverflow.com/questions/28608385/assert-asserting-when-debug-false
https://stackoverflow.com/questions/56990925/is-this-assert-for-development-not-for-production


TLDR:

* use exceptions for the user-side or external system-side of the package, where things can go wrong by unintended input/output, or maybe errors in servers, sockets or filesystem (like an OSError will be raised when you try to download but there's no space left on your device)
* use assertions for the developer-side of the package, to assert conditions that should absolutely never go wrong, independent of user input. in production this is more like checking your loop-invariants because you're paranoid. otherwise, asserts should be exclusively used in testing code.